Mathlib/Analysis/Calculus/FDeriv/Mul.lean

```diff
@@ -566,6 +566,229 @@ theorem fderiv_const_mul (ha : DifferentiableAt 𝕜 a x) (b : 𝔸) :
+theorem hasStrictFDerivAt_finset_prod [DecidableEq ι] [Fintype ι] {x : ι → 𝔸'} :
+    HasStrictFDerivAt (𝕜 := 𝕜) (∏ i in u, · i) (∑ i in u, (∏ j in u.erase i, x j) • proj i) x := by
+  simp only [Finset.sum_eq_multiset_sum, Finset.prod_eq_multiset_prod]
+  exact hasStrictFDerivAt_multiset_prod
+
+theorem hasFDerivAt_finset_prod [DecidableEq ι] [Fintype ι] {x : ι → 𝔸'} :
+    HasFDerivAt (𝕜 := 𝕜) (∏ i in u, · i) (∑ i in u, (∏ j in u.erase i, x j) • proj i) x :=
+  hasStrictFDerivAt_finset_prod.hasFDerivAt
+
+section Comp
+
+theorem HasStrictFDerivAt.list_prod' {l : List ι} {x : E}
+    (h : ∀ i ∈ l, HasStrictFDerivAt (f i ·) (f' i) x) :
+    HasStrictFDerivAt (fun x ↦ (l.map (f · x)).prod)
+      (∑ i : Fin l.length, ((l.take i).map (f · x)).prod •
+        smulRight (f' (l.get i)) ((l.drop (.succ i)).map (f · x)).prod) x := by
+  simp only [← List.finRange_map_get l, List.map_map]
+  refine .congr_fderiv (hasStrictFDerivAt_list_prod_finRange'.comp x
+    (hasStrictFDerivAt_pi.mpr fun i ↦ h (l.get i) (l.get_mem i i.isLt))) ?_
+  ext m
+  simp [← Function.comp_def (f · x) (l.get ·), ← List.map_map, List.map_take, List.map_drop]
+
+/--
+Unlike `HasFDerivAt.finset_prod`, supports non-commutative multiply and duplicate elements.
+-/
+theorem HasFDerivAt.list_prod' {l : List ι} {x : E}
+    (h : ∀ i ∈ l, HasFDerivAt (f i ·) (f' i) x) :
+    HasFDerivAt (fun x ↦ (l.map (f · x)).prod)
+      (∑ i : Fin l.length, ((l.take i).map (f · x)).prod •
+        smulRight (f' (l.get i)) ((l.drop (.succ i)).map (f · x)).prod) x := by
+  simp only [← List.finRange_map_get l, List.map_map]
+  refine .congr_fderiv (hasFDerivAt_list_prod_finRange'.comp x
+    (hasFDerivAt_pi.mpr fun i ↦ h (l.get i) (l.get_mem i i.isLt))) ?_
+  ext m
+  simp [← Function.comp_def (f · x) (l.get ·), ← List.map_map, List.map_take, List.map_drop]
+
+theorem HasFDerivWithinAt.list_prod' {l : List ι} {x : E}
+    (h : ∀ i ∈ l, HasFDerivWithinAt (f i ·) (f' i) s x) :
+    HasFDerivWithinAt (fun x ↦ (l.map (f · x)).prod)
+      (∑ i : Fin l.length, ((l.take i).map (f · x)).prod •
+        smulRight (f' (l.get i)) ((l.drop (.succ i)).map (f · x)).prod) s x := by
+  simp only [← List.finRange_map_get l, List.map_map]
+  refine .congr_fderiv (hasFDerivAt_list_prod_finRange'.comp_hasFDerivWithinAt x
+    (hasFDerivWithinAt_pi.mpr fun i ↦ h (l.get i) (l.get_mem i i.isLt))) ?_
+  ext m
+  simp [← Function.comp_def (f · x) (l.get ·), ← List.map_map, List.map_take, List.map_drop]
+
+theorem fderiv_list_prod' {l : List ι} {x : E}
+    (h : ∀ i ∈ l, DifferentiableAt 𝕜 (f i ·) x) :
+    fderiv 𝕜 (fun x ↦ (l.map (f · x)).prod) x =
+      ∑ i : Fin l.length, ((l.take i).map (f · x)).prod •
+        smulRight (fderiv 𝕜 (fun x ↦ f (l.get i) x) x)
+          ((l.drop (.succ i)).map (f · x)).prod :=
+  (HasFDerivAt.list_prod' fun i hi ↦ (h i hi).hasFDerivAt).fderiv
+
+theorem fderivWithin_list_prod' {l : List ι} {x : E}
+    (h : ∀ i ∈ l, DifferentiableWithinAt 𝕜 (f i ·) s x) (hxs : UniqueDiffWithinAt 𝕜 s x) :
```

Comment:
```suggestion
    (hxs : UniqueDiffWithinAt 𝕜 s x) (h : ∀ i ∈ l, DifferentiableWithinAt 𝕜 (f i ·) s x) :
```

Reply:
This seems to be the consistent argument order (at least in this file).

Reply:
Nice catch, thanks